### PR TITLE
Refactor Glass Recycler to use TGUI

### PR DESCRIPTION
**Pull request by blackrep**
[QOL][TGUI]

## About the PR 
This PR does two-fold:
1. datumizes the glass recycler's option list making it easier to read / edit
2. converts the glass recycler's UI to TGUI

## Why's this needed? 
TGUI is much prettier compared to the old UI.
screm

## Changelog

```changelog
(u)Blackrep
(+)The glass recycler has been updated to a fancy new interface!
```

### CODE REVIEW

**Comment by blackrep** [comment]
Would love to get early feedback on the datumization of the inventory list before I proceed with implementing the TGUI.

File: code/obj/machinery/glass_recycler.dm [hunk 1]
```diff
@@ -1,5 +1,34 @@
 #define PLATE_COST 2
 
+/datum/data/glass_product
```
Comment (blackrep) [inline code/obj/machinery/glass_recycler.dm:3]: This probably doesn't need to be a child of data, but was borrowing ideas from GardenGear.

File: tgui/packages/tgui/interfaces/GlassRecycler.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
+import { Fragment } from 'inferno';
+import { useBackend, useLocalState } from "../backend";
+import { Box, Button, Divider, Flex, Section, Stack } from "../components";
+import { Window } from "../layouts";
+
+// This already exists in WeaponVendor/index.tsx, but feels weird
+// to make it a dependency.
+const pluralize = (word, n) => (n !== 1 ? word + 's' : word);
+const capitalize = (word) => word.replace(/(^\w{1})|(\s+\w{1})/g, letter => letter.toUpperCase());
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/GlassRecycler.js:9]: should be abstracted out to a helper file for both then

Comment (blackrep) [reply tgui/packages/tgui/interfaces/GlassRecycler.js:9]: There a specific place I should be making this util file? Is placing it in `/tgui/packages/tgui` fine?

Comment (blackrep) [reply tgui/packages/tgui/interfaces/GlassRecycler.js]: Refactored into helper file `stringUtils`.

File: code/obj/machinery/glass_recycler.dm [hunk 3]
```diff
@@ -76,181 +118,100 @@
...
+		product_list += new /datum/glass_product("pitcher", /obj/item/reagent_containers/food/drinks/drinkingglass/pitcher, 2)
+
+	proc/create(var/type)
+		var/datum/glass_product/target_product = null
+		for (var/datum/glass_product/product in product_list)
+			if(product.product_type == type)
+				target_product = product
+				break
+
+		if (target_product == null)
```
Comment (Zonespace27) [inline code/obj/machinery/glass_recycler.dm:156]:
```suggestion
		if (!target_product)
```

File: code/obj/machinery/glass_recycler.dm [hunk 4]
```diff
@@ -76,181 +118,100 @@
...
-					<A href='?src=\ref[src];type=bottle'>Bottle</A><br>
-					<A href='?src=\ref[src];type=vial'>Vial</A><br>
-					<A href='?src=\ref[src];type=flask'>Flask</A>
-					<HR><A href='?src=\ref[src];refresh=1'>Refresh</A>
-					<BR><BR><A href='?action=mach_close&window=glass'>Close</A>"}
-		//user << browse(dat, "window=glass;size=220x240")
-		user.Browse(dat, "window=glass;size=220x360;title=Recycler")
-		onclose(user, "glass")
+		src.visible_message("<span class='notice'>[src] manufactures \a [G]!</span>")
 		return
```
Comment (Zonespace27) [inline code/obj/machinery/glass_recycler.dm:224]: Unneeded return

File: code/obj/machinery/glass_recycler.dm (hunk 3 above)
Comment (blackrep) [reply code/obj/machinery/glass_recycler.dm]: Implemented.

File: code/obj/machinery/glass_recycler.dm (hunk 4 above)
Comment (blackrep) [reply code/obj/machinery/glass_recycler.dm:224]: Removed.

**Comment by blackrep** [comment]
Side note: One thing I really wanted to add was categories, to break down the long list of items into specific tabs. However I couldn't think out a mapping of item to category that made it seem like it'd be worth it.

Was originally thinking Kitchenware, and Chemistryware. But chemistryware would likely only have 4 options in it (Beaker, Large Beaker, Vial, Bottle).

If there's still interest in categories I can take another crack at it.

File: tgui/packages/tgui/interfaces/GlassRecycler.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    disabled,
+    onClick,
+  } = props;
+
+  return (
+    <Fragment>
+      <Flex direction="row" align="center">
+        <Flex.Item>
+          <img
+            src={`data:image/jpeg;base64,${img}`}
```
Comment (ZeWaka) [inline tgui/packages/tgui/interfaces/GlassRecycler.js:24]: not `image/png` ? That's what DMIs literally are

Comment (ZeWaka) [reply tgui/packages/tgui/interfaces/GlassRecycler.js:24]:
```suggestion
            src={`data:image/png;base64,${img}`}
```

File: code/obj/machinery/glass_recycler.dm [hunk 6]
```diff
@@ -76,181 +118,99 @@
...
+		)
+
+	ui_act(action, params)
+		. = ..()
+		if (.)
 			return
+		switch(action)
+			if("create")
+				var/product_type = params["type"]
+				create(product_type)
```
Comment (ZeWaka) [inline code/obj/machinery/glass_recycler.dm:202]: review note: this function validates the input and doesn't just make the glass

**Review by ZeWaka — CHANGES_REQUESTED** [review]

**Review by ZeWaka — APPROVED** [review]

**Comment by ZeWaka** [comment]
Next time, you need to push the compiled bundled tgui changes. There is a bug in the GitHub checks that doesn't catch this currently.

**Comment by blackrep** [comment]
> Next time, you need to push the compiled bundled tgui changes. There is a bug in the GitHub checks that doesn't catch this currently.

Yep, was just talking with Pali about this on Discord. Will remember to do that for future PRs.

**Pull request merged**